subprojects/base-services/src/main/java/org/gradle/internal/operations/DefaultBuildOperationQueue.java

```diff
@@ -129,20 +148,52 @@ private static String getFailureMessage(Collection<? extends Throwable> failures
         return "Multiple build operations failed.";
     }
 
+    private class SubmittedWorker {
```

Comment:
We don't really need to restructure the workers. Instead, when waiting, increment `workers` before calling `run()` on the `WorkerRunnable`.